packages/@react-stately/dnd/src/useDroppableCollectionState.ts

```diff
@@ -29,38 +30,93 @@ export interface DroppableCollectionState {
 }
 
 export function useDroppableCollectionState(props: DroppableCollectionStateOptions): DroppableCollectionState  {
+  let {
+    acceptedDragTypes = 'all',
+    onInsert,
+    onRootDrop,
+    onItemDrop,
+    onReorder,
+    isValidDropTarget,
+    collection,
+    selectionManager,
+    onDropExit,
+    onDropEnter,
+    getDropOperation,
+    onDrop
+  } = props;
   let [target, setTarget] = useState<DropTarget>(null);
 
   let getOppositeTarget = (target: ItemDropTarget): ItemDropTarget => {
     if (target.dropPosition === 'before') {
-      let key = props.collection.getKeyBefore(target.key);
+      let key = collection.getKeyBefore(target.key);
       return key != null ? {type: 'item', key, dropPosition: 'after'} : null;
     } else if (target.dropPosition === 'after') {
-      let key = props.collection.getKeyAfter(target.key);
+      let key = collection.getKeyAfter(target.key);
       return key != null ? {type: 'item', key, dropPosition: 'before'} : null;
     }
   };
 
+  let defaultGetDropOperation = useCallback((target: DropTarget, types: DragTypes, allowedOperations: DropOperation[]) => {
+    // TODO: will be covered in a separate PR (issue in board), for now ignore the typescript error
+    // @ts-ignore
+    let typesSet = types.types ? types.types : types;
+    let draggedTypes = [...typesSet.values()];
+    let {draggingCollectionRef, currentDropCollectionRef, draggingKeys} = getDnDState();
+    // Compare draggingCollectionRef to currentDropCollectionRef since droppedCollectionRef hasn't been set yet (drop hasn't happened)
+    let isInternalDrop = draggingCollectionRef?.current === currentDropCollectionRef?.current;
+
+    if (
+      (acceptedDragTypes === 'all' || draggedTypes.every(type => acceptedDragTypes.includes(type))) &&
+      (
+        (onInsert && target.type === 'item' && (!isInternalDrop || allowedOperations[0] === 'copy') && (target.dropPosition === 'before' || target.dropPosition === 'after')) ||
+        (onReorder && target.type === 'item' && isInternalDrop && (target.dropPosition === 'before' || target.dropPosition === 'after')) ||
+        // Feedback was that internal root drop was weird so preventing that from happening
+        (onRootDrop && target.type === 'root' && !isInternalDrop) ||
+        // Automatically prevent items (i.e. folders) from being dropped on themselves.
+        // TODO: this also prevents non-folder items from being dropped on themseleves as well, is that too restrictive? Can't really think of a reason to allow that
+        (onItemDrop && target.type === 'item' && target.dropPosition === 'on' && !(isInternalDrop && draggingKeys.has(target.key)) && (!isValidDropTarget || isValidDropTarget(target, types))) ||
```

Comment:
Also note the user provided `isValidDropTarget` is only called for onItemDrop cases. I figured that the dynamic drop target filtering that this utility function provides is really only important for "on item" drop cases (e.g. disabling drops on a folder because the end user doesn't have write permissions). I don't think it will come into play for insertion or root drop operations since the hook user could omit `onInsert`/`onReorder`/`onRootDrop` entirely instead but open to opinions here